primitives/trie/src/trie_node.rs

```diff
@@ -253,100 +253,108 @@ impl<A: Serialize + Deserialize + Clone> TrieNode<A> {
+            hasher.finish()
+        })
+    }
 
-        size
+    pub fn hash_assert<H: HashOutput>(&self) -> H {
+        self.hash()
+            .expect("can only hash TrieNode with complete information about children")
     }
 }
 
-impl<A: Serialize + Deserialize + Clone> Deserialize for TrieNode<A> {
+impl Deserialize for TrieNode {
     fn deserialize<R: ReadBytesExt>(reader: &mut R) -> Result<Self, SerializingError> {
-        let node_type: TrieNodeType = Deserialize::deserialize(reader)?;
-
-        let prefix: KeyNibbles = Deserialize::deserialize(reader)?;
-
-        match node_type {
-            TrieNodeType::LeafNode => {
-                let value: A = Deserialize::deserialize(reader)?;
-
-                Ok(TrieNode::new_leaf(prefix, value))
-            }
-            TrieNodeType::BranchNode => {
-                let child_count: u8 = Deserialize::deserialize(reader)?;
-
-                let mut children = NO_CHILDREN;
-
-                for _ in 0..child_count {
-                    let child: TrieNodeChild = Deserialize::deserialize(reader)?;
+        let flags = reader.read_u8()?;
+        let has_root_data = flags & 0x1 != 0;
+        let has_value = flags & 0x2 != 0;
+        let has_children = flags & 0x4 != 0;
+        let root_data = if has_root_data {
+            Some(Deserialize::deserialize(reader)?)
+        } else {
+            None
+        };
+        let value = if has_value {
+            Some(DeserializeWithLength::deserialize::<u16, _>(reader)?)
+        } else {
+            None
+        };
+        let children = if has_children {
+            Deserialize::deserialize(reader)?
+        } else {
+            NO_CHILDREN
+        };
+        Ok(TrieNode {
+            // Make it clear that the key needs to be changed after deserialization.
+            key: KeyNibbles::BADBADBAD,
+            root_data,
+            value,
+            children,
+        })
+    }
+}
 
-                    if let Some(i) = child.suffix.get(0) {
-                        children[i] = Some(child);
-                    } else {
-                        return Err(io::Error::from(io::ErrorKind::InvalidData).into());
-                    }
-                }
+impl Serialize for TrieNode {
+    fn serialize<W: WriteBytesExt>(&self, writer: &mut W) -> Result<usize, SerializingError> {
+        let has_root_data = self.root_data.is_some();
+        let has_value = self.value.is_some();
+        let has_children = self.has_children();
+        let flags = (has_root_data as u8) | ((has_value as u8) << 1) | ((has_children as u8) << 2);
+
+        let mut result = 1;
+        writer.write_u8(flags)?;
+        if let Some(root_data) = &self.root_data {
+            result += root_data.serialize(writer)?;
+        }
+        if let Some(value) = &self.value {
+            result += value.serialize::<u16, _>(writer)?;
+        }
+        if has_children {
+            result += self.children.serialize(writer)?;
```

Comment:
This is not yet serializing the children into the compact format that we had before (only serialize children that are present, recover their position in the array from the child's prefix). This was a left-over tasks from last hackathon. Might be at odds now with children that use an empty prefix as a marker?